pkg/iampostgres/iampostgres.go

```diff
@@ -0,0 +1,159 @@
+import (
+	"database/sql"
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/jmoiron/sqlx"
+	"go.uber.org/zap"
+
+	pg "github.com/lib/pq"
+)
+
+type config struct {
+	useIAM           bool
+	passHolder       string
+	currentIamPass   string
+	currentPassMutex sync.Mutex
+	logger           Logger
+}
+
+var iamConfig = config{false, "", "", sync.Mutex{}, nil}
+
+// RDSPostgresDriver wrapper around postgres driver
+type RDSPostgresDriver struct {
+	*pg.Driver
+}
+
+// CustomPostgres is used to set the driverName to the custom postgres driver
+const CustomPostgres string = "custompostgres"
+
+// GetCurrentPass gets IAM password if needed and will block till valid password is available
+func GetCurrentPass() string {
+	// Blocks until the password from the dbConnectionDetails has a non blank password
+	currentPass := ""
+
+	counter := 0
+	maxCount := 120 // pauses for 30s
+
+	for {
+		counter++
+
+		iamConfig.currentPassMutex.Lock()
+		currentPass = iamConfig.currentIamPass
+		iamConfig.currentPassMutex.Unlock()
+
+		if currentPass == "" {
+			iamConfig.logger.Info(fmt.Sprintf("Wait %d of %d, sleeping for 250ms for IAM loop to populate RDS credentials.", counter, maxCount))
+		} else {
+			break
+		}
+		if counter > maxCount {
+			iamConfig.logger.Error("Waited 30s for IAM creds to populate and giving up, returning empty password.")
+			break
+		}
+
+		time.Sleep(time.Millisecond * 250)
+
+	}
+
+	return currentPass
+}
+
+func updateDSN(dsn string) (string, error) {
+	if !strings.Contains(dsn, iamConfig.passHolder) {
+		return "", errors.New("DSN does not contain password holder")
+	}
+
+	dsn = strings.Replace(dsn, iamConfig.passHolder, GetCurrentPass(), 1)
+	return dsn, nil
+}
+
+// Refreshes the RDS IAM on the given interval.
+func refreshRDSIAM(host string, port string, region string, user string, creds *credentials.Credentials, rus RDSUtilService, ticker *time.Ticker, logger Logger, errorMessagesChan chan error, shouldQuitChan chan bool) {
+	logger.Info("Starting refresh of RDS IAM")
+	// This for loop immediately runs the first tick then on interval
+	// This for loop will run indefinitely until it either errors or true is
+	// passed to the should quit channel.
+	for {
+		select {
+		case <-shouldQuitChan:
+			close(errorMessagesChan)
+			return
+		default:
+			if creds == nil {
+				logger.Error("IAM Credentials are missing")
+				errorMessagesChan <- errors.New("IAM Credientials are missing")
+				close(errorMessagesChan)
+				return
+			}
+			logger.Info("Using IAM Authentication")
+			authToken, err := rus.GetToken(host+":"+port, region, user, creds)
+			if err != nil {
+				logger.Error("Error building auth token", zap.Error(err))
+				errorMessagesChan <- fmt.Errorf("Error building auth token %v", err)
+				close(errorMessagesChan)
+				return
+			}
+
+			iamConfig.currentPassMutex.Lock()
+			iamConfig.currentIamPass = url.QueryEscape(authToken)
```

Comment:
```suggestion
			iamConfig.currentIamPass = authToken
```
Something to try if there are still issues when trying to use GetCurrentPass().